### PR TITLE
Contracts and cps call

### PR DESCRIPTION
**Pull request by partamonov**
List call for CPS service and All Contracts service calls

### CODE REVIEW

File: service/contractsv1/api.go [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+			return
+		}
+	}
+	return
+}
+
+// ListContracts gets the list of contracts that a user has access to.
+// 'depth' returns a specific set of contracts.
+// Select TOP to return only parent contracts or ALL to return both parent and child contracts.
+func (c *Contractsv1) ListContracts(depth string) (*OutputContractIDs, error) {
```
Comment (RafPe) [inline service/contractsv1/api.go]: Cannot this be done in the same way we have network lists ? Or with constants specific types ?

File: service/contractsv1/api.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+// ListProductsPerContract gets the IDs and names of the products associated with a contract for the time frame selected.
+// From - Ex: 2015-01-31. The start date, in UTC, to use when looking for products associated with a contract.
+//        The search always begins at midnight (0:00) UTC of the specified date.
+//        The default start date is 30 days prior to the current date.
+//        For current contracts, you can select a date within the past 15 months of the current date.
+//        For expired contracts, you are limited to a date range of 30 days within the 15 month window.
+// To - Ex: 2016-03-31. The end date, in UTC, to use when looking for products associated with a contract.
+//      The search always ends at 23:59:59 UTC of the specified date. The default end date is the current date.
+func (c *Contractsv1) ListProductsPerContract(contractID, from, to string) (*OutputContracts, error) {
```
Comment (RafPe) [inline service/contractsv1/api.go:44]: Use maybe timespan ?

File: service/contractsv1/api.go [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+	return resp.Result().(*OutputReportingGroupIDs), nil
+}
+
+// ListProductsPerReportingGroup gets the IDs and names of the products associated with the reporting group for the time frame selected.
+// To run this operation, your user account needs the CPCode Rep Group role.
+// When a request is successful, it may return either a 200 or a 300 response.
+// The API returns a 200 (OK) response when the CP code reporting group is associated with only one contract.
+// You’ll receive a 300 (Multiple Choices) response when the request returns a list of matching contracts because
+// the CP code reporting group is associated with multiple contracts.
+// To retrieve product information when you receive a 300 response code, make a new GET request to the hyperlinks provided in the response.
```
Comment (RafPe) [inline service/contractsv1/api.go:134]: 🤔 We need to make request towards response hyperlinks ?

File: service/cpsv2/types.go [hunk 4]
```diff
@@ -0,0 +1,211 @@
+package cpsv2
+
+// OutputEnrollments identifies a collection of OutputEnrollmentElement objects
+type OutputEnrollments struct {
+	Enrollments []OutputEnrollmentElement `json:"enrollments"`
+}
+
+// OutputEnrollmentElement displays all the information about the process that your certificate goes through
```
Comment (RafPe) [inline service/cpsv2/types.go:8]: Whats the purpose of `Output` keyword if there is no Input object ? Would it not make sense to unify this a little bit ?

**Review by RafPe — CHANGES_REQUESTED** [review]
Check if any of the comments would make sense to change in the end release

File: service/contractsv1/api.go (hunk 3 above)
Comment (partamonov) [reply service/contractsv1/api.go:134]: Yes, if you will get 300

File: service/cpsv2/types.go (hunk 4 above)
Comment (partamonov) [reply service/cpsv2/types.go:8]: output is returned from server, if I will later implement creation part there will be body objects which might be different based on exp with other APIs

File: service/contractsv1/api.go (hunk 2 above)
Comment (partamonov) [reply service/contractsv1/api.go:44]: You can, but do not want to add complexity here, can a feature request later to validate that given string is in right format, but that will be done by Akamai as well

File: service/contractsv1/api.go (hunk 1 above)
Comment (partamonov) [reply service/contractsv1/api.go]: Done

File: service/contractsv1/api.go (hunk 3 above)
Comment (RafPe) [reply service/contractsv1/api.go:134]: I think it is  something we should not be doing in API. IMHO if  that happens a result of specific type should be returned specific to this particular situation with a flag indicating a different action should be taken by hugher level api

File: service/cpsv2/types.go (hunk 4 above)
Comment (RafPe) [reply service/cpsv2/types.go:8]: But thats like tryiing to solve something we dont face yet. if we are using Output objects we indicate we have input - which is false in this scenario

Comment (partamonov) [reply service/cpsv2/types.go:8]: Api itself has input objects and not to rename later, as I might implement them for TF, I would keep output

File: service/contractsv1/api.go (hunk 3 above)
Comment (partamonov) [reply service/contractsv1/api.go:134]: This is why I return 2 values of certain types and error. And based on result you can decide what to do in higher level API.
```
products, contracts, err := api.ListProductsPerReportingGroup(ID, "", "")

if contracts != nil {...}
```

Comment (RafPe) [reply service/contractsv1/api.go:134]: The root thing here is `make a new GET request to the hyperlinks provided in the response` which I would not to expect in API system.
So based on the above must I do `contracts[0].Hyperlink` and GET ?

File: service/cpsv2/types.go (hunk 4 above)
Comment (RafPe) [reply service/cpsv2/types.go:8]: Ok - makes sense

File: service/contractsv1/api.go (hunk 3 above)
Comment (partamonov) [reply service/contractsv1/api.go:134]: But here you are trying to solve design issue of API implementation. I agree that is odd.
But rephrasing what is written in docs. If you want to list products associated with reporting group, but group itself is associated with more then 1 contract you cannot do that and you will get back a list of contracts ID back with href. And to get product detail you have to make calls in for cycle ListProductsPerContract(ID)
```
for _ , c := range contracts {
  p := api.ListProductsPerContract(c.ContractID)
}
```

Comment (RafPe) [reply service/contractsv1/api.go:134]: Ok - makes sense - then just for our awesomness ;) change the description as it implies we implemented API in very strange way

**Review by RafPe — APPROVED** [review]

**Pull request merged**